File: OpenInPlace/Working Copy/WorkingCopyUrlService.m
//
//  WorkingCopyGitService.m
//  WorkingCopy
//
//  Created by Anders Borum on 13/08/2018.
//  Copyright © 2018 Applied Phasor. All rights reserved.
//

#import "WorkingCopyUrlService.h"

#define ServiceNameVer1 @"working-copy-v1"
#define ServiceNameVer2 @"working-copy-v2"

@protocol WorkingCopyProtocolVer1

-(void)determineDeepLinkWithCompletionHandler:(void (^)(NSURL* url))completionHandler;

-(void)fetchDocumentSourceInfoWithCompletionHandler:(void (^)(NSString* path,
                                                              NSString* appName,
                                                              NSString* appVersion,
                                                              NSData* appIconPNG))completionHandler;

@end

@protocol WorkingCopyProtocolVer2 <WorkingCopyProtocolVer1>

-(void)fetchStatusAvatarLength:(NSUInteger)length // 0 to not get back avatar
             completionHandler:(void (^)(NSUInteger linesAdded,
                                         NSUInteger linesDeleted,
                                         NSString* commitIdentifier,
                                         NSString* commitAuthor,
                                         NSData* commitAvatarPng,
                                         NSDate* commitDate,
                                         NSError* error))completionHandler;

-(void)fetchBlameAtLocation:(NSUInteger)location // NSNotFound to not check blame for location
               avatarLength:(NSUInteger)length // 0 to not get back avatar
          completionHandler:(void (^)(NSString* commitIdentifier,
                                      NSString* commitAuthor,
                                      NSData* commitAvatarPng,
                                      NSDate* commitDate,
                                      NSError* error))completionHandler;

@end

@interface WorkingCopyUrlService () {
    NSXPCConnection* connection;
    id<WorkingCopyProtocolVer1> proxy1;
    id<WorkingCopyProtocolVer2> proxy2;

    NSError* error;
    void (^errorHandler)(NSError* error);
}

@end

@implementation WorkingCopyUrlService

-(void)determineDeepLinkWithCompletionHandler:(void (^_Nonnull)(NSURL* _Nullable url,
                                                                NSError* _Nullable error))completionHandler {
    errorHandler = ^(NSError* error) {
        completionHandler(nil, error);
    };
    
    [proxy1 determineDeepLinkWithCompletionHandler:^(NSURL* url) {
        NSError* theError = [self->error copy];

        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(url, theError);
        });
    }];
}

-(void)fetchDocumentSourceInfoWithCompletionHandler:(void (^_Nonnull)(NSString* _Nullable path,
                                                                      NSString* _Nullable appName,
                                                                      NSString* _Nullable appVersion,
                                                                      UIImage* _Nullable appIcon,
                                                                      NSError* _Nullable error))completionHandler {
    errorHandler = ^(NSError* error) {
        completionHandler(nil, nil, nil, nil, error);
    };
    
    [proxy1 fetchDocumentSourceInfoWithCompletionHandler:^(NSString* path,
                                                          NSString* appName,
                                                          NSString* appVersion,
                                                          NSData* iconPNG) {
        NSError* theError = [self->error copy];
        UIImage* icon = iconPNG == nil ? nil : [UIImage imageWithData:iconPNG];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(path, appName, appVersion, icon, theError);
        });
    }];
}

-(void)fetchStatusWithCompletionHandler:(void (^_Nonnull)(NSUInteger linesAdded, NSUInteger linesDeleted,
                                                          NSError* _Nullable error))completionHandler {
    if(proxy2 == nil) {
        NSString* message = NSLocalizedString(@"Status check requires Working Copy 3.5.0 or later.", nil);
        NSDictionary* userInfo = @{NSLocalizedDescriptionKey: message};
        NSError* error = [NSError errorWithDomain:@"Working Copy" code:400 userInfo:userInfo];
        completionHandler(0,0, error);
        return;
    }
    
    errorHandler = ^(NSError* error) {
        completionHandler(0,0, error);
    };
    
    [proxy2 fetchStatusAvatarLength:100
                  completionHandler:^(NSUInteger linesAdded,
                                      NSUInteger linesDeleted,
                                      NSString* commitIdentifier,
                                      NSString* commitAuthor,
                                      NSData* commitAvatarPng,
                                      NSDate* commitDate,
                                      NSError* error) {
                    
        NSError* theError = error ?: [self->error copy];
                      
        UIImage* avatar = commitAvatarPng == nil ? nil : [UIImage imageWithData:commitAvatarPng];
                    
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(linesAdded, linesDeleted, theError);
        });
    }];
}

-(instancetype)initWithConnection:(NSXPCConnection*)theConnection
                      serviceName:(NSString*)serviceName {
    self = [super init];
    if(self != nil) {
        connection = theConnection;

        Protocol* protocol = nil;
        if([serviceName isEqualToString:ServiceNameVer2]) {
            protocol = @protocol(WorkingCopyProtocolVer2);
        } else {
            protocol = @protocol(WorkingCopyProtocolVer1);
        }
        
        connection.remoteObjectInterface = [NSXPCInterface interfaceWithProtocol:protocol];
        [connection resume];
        
        proxy1 = [connection remoteObjectProxyWithErrorHandler:^(NSError* theError) {
            self->error = theError;
            [self->connection invalidate];
            
            if(self->errorHandler) {
                // make sure error handler is only called once
                void (^copy)(NSError* error) = [self->errorHandler copy];
                self->errorHandler = nil;
                dispatch_async(dispatch_get_main_queue(), ^{
                    copy(theError);
                });
            }
        }];
        
        if([serviceName isEqualToString:ServiceNameVer2]) {
            proxy2 = (id<WorkingCopyProtocolVer2>)proxy1;
        }
    }
    return self;
}

-(void)dealloc {
    [connection invalidate];
}

+(void)getServiceForUrl:(nonnull NSURL*)url
      completionHandler:(void (^_Nonnull)(WorkingCopyUrlService* _Nullable service,
                                          NSError* _Nullable error))completionHandler {
    
    BOOL securityScoped = [url startAccessingSecurityScopedResource];
    
    [[NSFileManager defaultManager] getFileProviderServicesForItemAtURL:url
                                                      completionHandler:^(NSDictionary* services,
                                                                          NSError* error) {
          // check that we have provider service
          NSFileProviderService* providerService = services[ServiceNameVer2];
          if(providerService == nil) providerService = services[ServiceNameVer1];
                                                          
          if(error != nil || providerService == nil) {
              dispatch_async(dispatch_get_main_queue(), ^{
                  completionHandler(nil, error);
              });
              if(securityScoped) {
                  [url stopAccessingSecurityScopedResource];
              }
              return;
          }
                                                          
          // attempt connection
          [providerService getFileProviderConnectionWithCompletionHandler:^(NSXPCConnection* connection,
                                                                            NSError* error) {
              
              if(securityScoped) {
                  [url stopAccessingSecurityScopedResource];
              }
              
              // make sure we have connection
              if(error != nil || connection == nil) {
                  dispatch_async(dispatch_get_main_queue(), ^{
                      completionHandler(nil, error);
                  });
                  return;
              }
             
              // setup proxy object
              WorkingCopyUrlService* service = [[WorkingCopyUrlService alloc] initWithConnection:connection
                                                                                     serviceName:providerService.name];
              dispatch_async(dispatch_get_main_queue(), ^{
                  completionHandler(service, nil);
              });
        }];
    }];
}

@end

File: OpenInPlace/Working Copy/WorkingCopyUrlService.swift
//
//  WorkingCopyUrlService.swift
//  OpenInPlace
//
//  Created by Anders Borum on 24/09/2018.
//  Copyright © 2018 Applied Phasor. All rights reserved.
//

import UIKit

@objc private protocol WorkingCopyProtocolVer1 {
    func determineDeepLinkWithCompletionHandler(_ completionHandler:
                                                @escaping ((URL?) -> Void))
    
    func fetchDocumentSourceInfoWithCompletionHandler(_ completionHandler:
                                                      @escaping ((String?, String?,
                                                                  String?, Data?) -> Void))
}

@objc private protocol WorkingCopyProtocolVer352 : WorkingCopyProtocolVer1 {
    func fetchStatusWithCompletionHandler(_ completionHandler:
                                          @escaping ((UInt, UInt, Error?) -> Void))
}

/// Retrieves information about files and directories stored in a Working Copy
/// file provider.
///
/// You initialise instances from a file URL that the user has granted your app
/// access using the document picker, document browser, drag and drop or by
/// opening a file in-place.
class WorkingCopyUrlService {
    private static let serviceNameVer1 = NSFileProviderServiceName("working-copy-v1")
    private static let serviceNameVer352 = NSFileProviderServiceName("working-copy-v3.5.2")
    
    private var connection: NSXPCConnection
    private var proxy1: WorkingCopyProtocolVer1?
    private var proxy352: WorkingCopyProtocolVer352?
    
    private var error: Error?
    private var errorHandler: ((Error) -> Void)?
    
    /// Try to inquire and connect to WorkingCopyUrlService on the given URL.
    /// 
    /// Note that you can get a nil-service even without a error when url is outside
    /// a Working Copy file provider.
    ///
    /// Completion handler is called on main thread.
    @available(iOS 11.0, *) class public func getFor(_ url: URL,
                                                     completionHandler: @escaping ((WorkingCopyUrlService?, Error?) -> ())) {
        let securityScoped = url.startAccessingSecurityScopedResource();
        FileManager.default.getFileProviderServicesForItem(at: url, completionHandler: {
            (services, error) in
            
            // check that we have provider service
            let potentialService = services?[serviceNameVer352] ?? services?[serviceNameVer1]
            
            guard let providerService = potentialService, error == nil else {
                DispatchQueue.main.async {
                    completionHandler(nil, error)
                }
                if securityScoped {
                    url.stopAccessingSecurityScopedResource()
                }
                return
            }
            
            // attempt connection
            providerService.getFileProviderConnection(completionHandler: {
                (connection, error) in
                
                if securityScoped {
                    url.stopAccessingSecurityScopedResource()
                }
                
                // make sure we have connection
                guard let theConnection = connection, error == nil else {
                    DispatchQueue.main.async {
                        completionHandler(nil, error)
                    }
                    return
                }
                
                // setup proxy object
                let service = WorkingCopyUrlService(theConnection, providerService.name)
                DispatchQueue.main.async {
                    completionHandler(service, nil)
                }
            })
        })
    }
    
    /// Determine deep-link for opening a the given URL inside Working Copy.
    ///
    /// This link is something on the form:
    ///
    ///     working-copy://open?repo=welcome%20to%20working%20copy&path=README.md
    ///
    /// Completion block is called on main thread.
    public func determineDeepLink(completionHandler: @escaping ((_ url: URL?,
                                                                 Error?) -> Void)) {
        
        errorHandler = { error in
            completionHandler(nil, error)
        }
        
        proxy1?.determineDeepLinkWithCompletionHandler({ url in
            let theError = self.error
            DispatchQueue.main.async {
                completionHandler(url, theError)
            }
        })
    }
    
    /// Determine path relative to Working Copy storage and app information
    /// that is shared by all Working Copy URLs.
    ///
    /// Completion block is called on main thread.
    public func fetchDocumentSourceInfo(completionHandler: @escaping ((_ path: String?,
                                                                       _ appName: String?,
                                                                       _ appVersion: String?,
                                                                       _ icon: UIImage?,
                                                                       Error?) -> Void)) {
        
        errorHandler = { error in
            completionHandler(nil, nil, nil, nil, error)
        }

        proxy1?.fetchDocumentSourceInfoWithCompletionHandler({
            (path, appName, appVersion, iconPNG) in
            
            let theError = self.error
            var icon: UIImage?
            if let png = iconPNG {
                icon = UIImage(data: png)
            }
            
            DispatchQueue.main.async {
                completionHandler(path, appName, appVersion, icon, theError)
            }
        })
    }
    
    /// Determine the lines added or deleted for the file at the given URL compared to last commit.
    /// If the file is current both lines added and deleted are zero, while NSNotFound indicates
    /// a modified binary file.
    ///
    /// Completion block is called on main thread.
    public func fetchStatus(completionHandler: @escaping ((_ linesAdded: UInt,
                                                           _ linesDeleted: UInt,
                                                           Error?) -> Void)) {
    
        guard let proxy = proxy352 else {
            let message = NSLocalizedString("Status check requires Working Copy 3.5.2 or later.",  comment: "")
            let userInfo = [NSLocalizedDescriptionKey: message]
            let error = NSError.init(domain: "Working Copy", code: 400, userInfo: userInfo)
            completionHandler(0,0, error)
            return
        }
        
        errorHandler = { error in
            completionHandler(0,0, error)
        }
    
        proxy.fetchStatusWithCompletionHandler({
            (linesAdded, linesDeleted, error) in
            
            let theError = error ?? self.error
            DispatchQueue.main.async {
                completionHandler(linesAdded, linesDeleted,
                                  theError)
            }
        })
    }
    
    private init(_ theConnection: NSXPCConnection,
                 _ serviceName: NSFileProviderServiceName) {
        connection = theConnection
        
        var theProtocol: Protocol?
        if serviceName == WorkingCopyUrlService.serviceNameVer352 {
            theProtocol = WorkingCopyProtocolVer352.self
        } else {
            theProtocol = WorkingCopyProtocolVer1.self
        }
        
        connection.remoteObjectInterface =  NSXPCInterface(with: theProtocol!)
        connection.resume()
        
        proxy1 = connection.remoteObjectProxyWithErrorHandler({
            error in
            
            self.error = error
            self.connection.invalidate()
            
            if let handler = self.errorHandler {
                // make sure error handler is only called once
                self.errorHandler = nil
                
                DispatchQueue.main.async {
                    handler(error)
                }
            }
        }) as? WorkingCopyProtocolVer1
        
        if serviceName == WorkingCopyUrlService.serviceNameVer352 {
            proxy352 = proxy1 as? WorkingCopyProtocolVer352
        }
        
    }
    
    deinit {
        connection.invalidate()
    }
}
